0010-B-object-storage-building-block.md

```diff
@@ -0,0 +1,275 @@
+  string key = 2;
+
+  // The tag for concurrency control.
+  string tag = 3;
+
+  // The metadata which will be stored with the object.
+  map<string, string> metadata = 4;
+
+  // Chunk of data.
+  StreamPayload payload = 10;
+}
+
+// SetObjectResponse contains the result of storing the object.
+message SetObjectResponse {
+  // The updated tag for concurrency control.
+  string tag = 1;
+
+  // The number of bytes written.
+  int bytes = 2;
+}
+
+// DeleteObjectRequest is the message to delete an object in specific object store.
+message DeleteObjectRequest {
+  // The name of object store.
+  string store_name = 1;
+
+  // The key of the desired object.
+  string key = 2;
+}
+
+// DeleteObjectResponse contains the result of deleting the object.
+message DeleteObjectResponse {
+    // Currently empty but allowing for future expansion.
+}
+```
+
+#### Handling streams
+
+`StreamPayload` is first introduced with dapr/dapr#4903 and corresponds to:
+
+```proto3
+// Chunk of data sent in a streaming request or response.
+message StreamPayload {
+  // Data sent in the chunk.
+  google.protobuf.Any data = 1;
+
+  // Set to true if this is the last chunk.
+  bool complete = 2;
+}
+```
+
+Data is sent from the application to Dapr (in `SetObject` RPCs) or from Dapr to the application (in `GetObject` RPCs) in a stream. Each message in the stream contains a `GetObjectResponse` or `SetObjectRequest` where:
+
+- The first message in the stream MUST contain all other required keys.
+- The first message in the stream MAY contain a `payload`, but that is not required.
+- Subsequent messages (any message except the first in the stream) MUST contain a `payload` and MUST NOT contain any other property.
+- The last message in the stream MUST contain a `payload` with `complete` set to `true`. That message is assumed to be the last one from the sender and no more messages are to be sent in the stream after that.
+
+The amount of data contained in `payload.data` is variable and it's up to the discretion of the sender. In service invocation calls, as implemented by dapr/dapr#4903, chunks are at most 4KB in size, although senders may send smaller chunks if they wish. Receivers must not assume that messages will contain any specific number of bytes in the payload.
+
+Note that it's possible for senders to send a single message in the stream. If the data is small and could fit in a single chunk, senders MAY choose to include a `payload` with `complete=true` in the first message. Receivers should assume that single message to be the entire communication from the sender.
+
+### Metadata and headers
+
+Metadata properties are passed between the client and server as headers.
+
+There can be two types of headers:
+
+- Custom ones have the `x-dapr-` prefix in requests sent between Dapr and the app. When they are stored, they generally do not have the `x-dapr-` prefix, which is removed (although this could be dependent on the implementation of the state store).
```

Comment:
IMHO, `x-dapr-` headers feel like they would be for out-of-band communication between the app and sidecar, and not for application metadata. If I recall correctly, most object stores will also send headers as-is from the object store when requested externally, thus potentially leaking information, such as the fact that dapr is being used (which may be undesirable, much like `wp-content` urls lets any attacker know that they're dealing with WordPress immediately). 

I don't have an answer here, just something to consider.